arbnode/espresso_watchdog.go

```diff
@@ -0,0 +1,244 @@
+		seqInbox:              seqInbox,
+		config:                config,
+		l1Reader:              l1Reader,
+		delayedMessageFetcher: delayedMessageFetcher,
+		fatalErrChan:          fatalErrChan,
+	}
+}
+
+func (f *ForceInclusionChecker) checkIfMessageCanBeForceIncluded(ctx context.Context) error {
+	// Get the total number of delayed messages read in the sequencer inbox
+	totalDelayedMessagesRead, err := f.seqInbox.TotalDelayedMessagesRead(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting total delayed messages read: %w", err)
+	}
+
+	// Get the earliest block number that is without the force inclusion tolerance
+	badBlockNumber := f.getForceInclusionToleranceBlockNumber(ctx)
+	// Check the delayed message count at this block number
+	count, err := f.delayedMessageFetcher.getDelayedMessageCountAtBlock(badBlockNumber)
+	if err != nil {
+		return fmt.Errorf("error getting delayed message count at block %d: %w", badBlockNumber, err)
+	}
+	// If the message count in delay inbox is less than or equal to the total delayed messages read
+	// then no force inclusion is going to happen.
+	if count <= arbmath.BigToUintSaturating(totalDelayedMessagesRead) {
+		return nil
+	}
+	// Force inclusion is going to happen, panic the node.
+	err = fmt.Errorf("force inclusion is going to happen")
+	f.fatalErrChan <- err
+	return err
+}
+
+func (f *ForceInclusionChecker) Start(ctx context.Context) error {
+	f.StopWaiter.Start(ctx, f)
+
+	return f.CallIterativelySafe(func(ctx context.Context) time.Duration {
+		err := f.checkIfMessageCanBeForceIncluded(ctx)
+		if err != nil {
+			log.Error("error checking force inclusion", "err", err)
+			return f.config.RetryTime
+		}
+		return f.config.PollingInterval
+	})
+}
+
+func (f *ForceInclusionChecker) getForceInclusionToleranceBlockNumber(ctx context.Context) uint64 {
+	maxTimeVariationDelayBlocks, _, maxTimeVariationDelaySeconds, _, err := f.seqInbox.MaxTimeVariation(ctx)
+	if err != nil {
+		return 0
+	}
+	currentParentChainBlock, err := f.l1Reader.Client().BlockByNumber(ctx, nil)
+	if err != nil {
+		return 0
+	}
+
+	lastBadBlockNumber := arbmath.SaturatingUSub(f.config.BlockThresholdTolerance+currentParentChainBlock.NumberU64(), arbmath.BigToUintSaturating(maxTimeVariationDelayBlocks))
+	lastBadBlockTime := arbmath.SaturatingUSub(f.config.SecondThresholdTolerance+currentParentChainBlock.Time(), arbmath.BigToUintSaturating(maxTimeVariationDelaySeconds))
+
+	lastBadBlock := f.findFirstParentChainBlockBelow(ctx, lastBadBlockNumber, lastBadBlockTime)
+	return lastBadBlock
+}
+
+func (f *ForceInclusionChecker) findFirstParentChainBlockBelow(ctx context.Context, lastBadBlockNumber uint64, lastBadBlockTime uint64) uint64 {
+	client := f.l1Reader.Client()
+	blockNumber := lastBadBlockNumber
+
+	for blockNumber > 0 {
+		block, err := client.BlockByNumber(ctx, arbmath.UintToBig(blockNumber))
+		if err != nil {
+			return 0
+		}
+		if block.NumberU64() <= lastBadBlockNumber || block.Time() <= lastBadBlockTime {
```

Comment:
should this not be `block.NumberU64() <= lastBadBlockNumber && block.Time() <= lastBadBlockTime`?

Reply:
Sorry for the late reply.

I think it should be `or` because that's[ what batcher does](https://github.com/EspressoSystems/nitro-espresso-integration/blob/161b3bbe3bf083f4ea4175c71d5ead858f4876bd/arbnode/batch_poster.go#L1639)

Reply:
@ImJeremyHe  I was reading arbitrum's code and it looks like they are doing something similar to && here
https://github.com/OffchainLabs/arbitrum-sdk/blob/792a7ee3ccf09842653bc49b771671706894cbb4/src/lib/inbox/inbox.ts#L123-L125

Let me know if that is not true!